backend.native/tests/runtime/workers/freeze6.kt

```diff
@@ -30,5 +34,125 @@ fun ensureNeverFrozenFailsTarget(){
     println("OK")
 }
 
-data class Hi(val s:String)
-data class Nested(val hi:Hi)
\ No newline at end of file
+fun createRef1(): FreezableAtomicReference<Any?> {
+    val ref = FreezableAtomicReference<Any?>(null)
+    ref.value = ref
+    ref.freeze()
+    ref.value = null
+
+    return ref
+}
+
+var global = 0
+
+@Test
+fun ensureFreezableHandlesCycles1() {
+    val ref = createRef1()
+    kotlin.native.internal.GC.collect()
+
+    val obj: Any = ref
+    global = obj.hashCode()
+}
+
+class Node(var ref: Any?)
+
+/**
+ *    ref1 -> Node <- ref3
+ *             |        /\
+ *             V        |
+ *             ref2  ---
+ */
+fun createRef2(): FreezableAtomicReference<Node?> {
+    val ref1 = FreezableAtomicReference<Node?>(null)
+
+    val node = Node(null)
+    val ref3 = FreezableAtomicReference<Any?>(node)
+    val ref2 = FreezableAtomicReference<Any?>(ref3)
+
+    node.ref = ref2
+    ref1.value = node
+
+    ref1.freeze()
+    ref3.value = null
+
+    assert(node.isFrozen)
+    assert(ref1.isFrozen)
+    assert(ref2.isFrozen)
+    assert(ref3.isFrozen)
+
+    return ref1
+}
+
+@Test
+fun ensureFreezableHandlesCycles2() {
+    val ref = createRef2()
+    kotlin.native.internal.GC.collect()
+
+    assert(ref.value.toString().length > 0)
+    global = ref.value!!.ref!!.hashCode()
+}
+
+fun createRef3(): FreezableAtomicReference<Any?> {
+    val ref = FreezableAtomicReference<Any?>(null)
+    val node = Node(ref)
+    ref.value = node
+    ref.freeze()
+
+    assert(node.isFrozen)
+    assert(ref.isFrozen)
+
+    return ref
+}
+
+@Test
+fun ensureFreezableHandlesCycles3() {
+    val ref = createRef3()
+    ref.value = null
+    kotlin.native.internal.GC.collect()
+
+    val obj: Any = ref
+    assert(obj.toString().length > 0)
+    global = obj.hashCode()
+}
+
+lateinit var weakRef: WeakReference<Any>
+
+fun createRef4(): FreezableAtomicReference<Any?> {
+    val ref = FreezableAtomicReference<Any?>(null)
+    val node = Node(ref)
+    weakRef = WeakReference(node)
+    ref.value = node
+    ref.freeze()
+    return ref
+}
+
+@Test
+fun ensureWeakRefNotLeaks1() {
+    val ref = createRef4()
+    assert(weakRef.get() != null)
+    ref.value = null
+    kotlin.native.internal.GC.collect()
+
+    assert(weakRef.get() == null)
```

Comment:
Note that assertions are disabled for regular tests currently. This `assert` actually fails.